### PR TITLE
Add tracking metric options; make metrics ModuleList; other improvements

### PR DESCRIPTION
**Pull request by KnathanM**
(Note: both #1022 and #1024 are based on this PR)

### tracking metric
Previously in the CLI the first metric in the list of metrics would be used for checkpointing and early stopping. This PR changes this to default to "val_loss" and give the user an option to change this.

I also changed how "val_loss" is calculated so that it is calculated the same as the training loss (using `train_step` instead of `forward`. The two values reported on the progress bar should be comparable. 
* The "val_loss" isn't exactly the same as the "train_loss" because during validation, batch norm and dropout are set to `eval()`. Also the "train_loss" shown is for a single batch, which may be much higher or lower than average. I change this to also calculate the train loss at the end of the epoch in #1022. 
* For some predictors, the output of `forward` and `train_step` are different, like whether a `sigmoid` is applied in `BinaryClassficationFFN`. Metrics are given the output of `forward` while loss functions are given the output of `train_step`, so `validation_step` has to run the data through the model twice, once through `forward` for metric calculations and once through `train_step` to calculate "val_loss". 
* Some predictors have the same metric class for both the default criterion and default metric, for example `RegressionFFN` has `MSE`. This means the MSE will be calculated twice for the validation set and reported as both "val_loss" and "val/mse". Maybe that could be optimized better.
* Also note that while the validation metric and test metric both use `_evaluate_batch`, the validation metric will be in the scaled space while the test metric will be in the original space, so they aren't directly comparable (except when the targets aren't scaled like in classification). Making them comparable would either require not scaling the validation targets, in which case "val_loss" and "training_loss" would not be comparable, or unscaling the validation metric value before logging, which seems complicated to get right for every metric. 

### metrics as a `ModuleList`
A separate issue this PR addresses is whether metrics should be registered as children of the `MPNN`. Maybe this should have been another PR.
* We have had issues with the metrics not being children of the `MPNN` class because then any of their parameters don't get moved between devices automatically. The simple solution is to make the list of metrics a `ModuleList` instead. 
* I had to add `eq=True, frozen=True` to `ThresholdedMixin` to have it be hashable so metrics that use it can be pickled.
* Another complication of metrics as a `ModuleList` is then any buffers of the metrics, i.e. the `task_weights` get saved to the state_dict. Old checkpoints are not compatible with this change because Chemprop will now expect there to be task weights for the (arbitrary number of) metrics in the state dict. 
* I had to manually load the checkpoint in `load_from_checkpoint` to add these task weights to the state dict. Then I save the model to a buffer to be loaded again  in the `lightning` framework.
* In #1024 I propose removing the task weights from the state dict instead.

While adding the task weights to the state dict, I went ahead and simplified loading the submodules (mp, agg, ffn) from `load_from_checkpoint` and `load_from_file` into one place.

### Other changes
Two other small changes:
* During evaluation we want the data point weights to be set to ones. Previously we used `weights = torch.ones_like(targets)`, but I think it makes more sense to get the (possibly not equal to 1) data weights from the batch and do `weights = torch.ones_like(weights)`. Then we won't have issues with multi-task targets and using `weights.view(-1,1)` in metrics 
* I make mask, weights, lt_weights, and gt_weights all optional, mostly because I got tired of having to make those in a notebook while testing the loss functions.

### CODE REVIEW

File: chemprop/cli/hpopt.py [hunk 1]
```diff
@@ -315,11 +315,17 @@ def train_model(config, args, train_dset, val_dset, logger, output_transform, in
...
-    monitor_mode = "min" if model.metrics[0].minimize else "max"
-    logger.debug(f"Evaluation metric: '{model.metrics[0].alias}', mode: '{monitor_mode}'")
+    if args.tracking_metric == "val_loss":
+        tracking_metric_class = model.criterion
+    else:
+        tracking_metric_class = MetricRegistry[args.tracking_metric]
+        args.tracking_metric = "val/" + args.tracking_metric
+
+    monitor_mode = "min" if tracking_metric_class.minimize else "max"
+    logger.debug(f"Evaluation metric: '{tracking_metric_class.alias}', mode: '{monitor_mode}'")
```
[the remark this answers is not in the document]
Comment (KnathanM) [reply chemprop/cli/hpopt.py:325]: I see now that `model.criterion` is the `tracking_metric`, but that `MetricRegistry[args.tracking_metric]` is the `tracking_metric_class`. I have changed the first one to also give the class for consistency. I don't need to use `args.tracking_metric.alias` because `args.tracking_metric` is already the string alias for the metric, which is why we can do `MetricRegistry[args.tracking_metric]`.

File: chemprop/cli/hpopt.py [hunk 2]
```diff
@@ -315,11 +315,17 @@ def train_model(config, args, train_dset, val_dset, logger, output_transform, in
     model = build_model(args, train_loader.dataset, output_transform, input_transforms)
     logger.info(model)
 
-    monitor_mode = "min" if model.metrics[0].minimize else "max"
-    logger.debug(f"Evaluation metric: '{model.metrics[0].alias}', mode: '{monitor_mode}'")
+    if args.tracking_metric == "val_loss":
+        tracking_metric_class = model.criterion
+    else:
+        tracking_metric_class = MetricRegistry[args.tracking_metric]
+        args.tracking_metric = "val/" + args.tracking_metric
```
[the remark this answers is not in the document]
Comment (KnathanM) [reply chemprop/cli/hpopt.py:322]: I'm more in favor of adding this check and then giving a `logger.warn` message if the metric/loss function doesn't match the predictor. I think this would be easiest to add after merging in the change to torchmetrics PR.

File: chemprop/cli/train.py [hunk 3]
```diff
@@ -864,15 +886,22 @@ def train_model(
 
         checkpointing = ModelCheckpoint(
             checkpoint_dir / "checkpoints",
-            "best-{epoch}-{val_loss:.2f}",
-            "val_loss",
+            "best-epoch={epoch}-"
```
[the remark this answers is not in the document]
Comment (KnathanM) [reply chemprop/cli/train.py:889]: Great idea

File: chemprop/models/model.py [hunk 4]
```diff
@@ -157,38 +158,43 @@ def on_validation_model_eval(self) -> None:
...
 
-        self.log_dict(metric2loss, batch_size=len(batch[0]))
+        bmg, V_d, X_d, targets, weights, lt_mask, gt_mask = batch
+
+        mask = targets.isfinite()
+        targets = targets.nan_to_num(nan=0.0)
+
+        Z = self.fingerprint(bmg, V_d, X_d)
+        preds = self.predictor.train_step(Z)
+        loss = self.metrics[-1](preds, targets, mask, weights, lt_mask, gt_mask)
```
[the remark this answers is not in the document]
Comment (KnathanM) [reply chemprop/models/model.py:170]: I see now this is a change that was supposed to go in the torchmetrics PR. In that PR, the last element in metrics is a copy of the criterion and not the criterion itself because the torchmetric has a saved state that can't be shared between training and validation. I could implement this change here, but would need to update the torchmetrics PR, so I would prefer no to.

File: chemprop/cli/train.py [hunk 5]
```diff
@@ -347,6 +347,11 @@ def add_train_args(parser: ArgumentParser) -> ArgumentParser:
         action=LookupAction(MetricRegistry),
         help="Specify the evaluation metrics. If unspecified, chemprop will use the following metrics for given dataset types: regression -> ``rmse``, classification -> ``roc``, multiclass -> ``ce`` ('cross entropy'), spectral -> ``sid``. If multiple metrics are provided, the 0-th one will be used for early stopping and checkpointing.",
     )
+    train_args.add_argument(
+        "--tracking-metric",
+        default="val_loss",
```
[the remark this answers is not in the document]
Comment (KnathanM) [reply chemprop/cli/train.py:352]: We can't use the lookup action because "val_loss" isn't in the MetricRegistry. Instead we handle the logic for changing this to an actual metric later.

File: chemprop/cli/train.py [hunk 6]
```diff
@@ -447,6 +452,13 @@ def add_train_args(parser: ArgumentParser) -> ArgumentParser:
 
 
 def process_train_args(args: Namespace) -> Namespace:
+    if args.output_dir is None:
+        args.output_dir = Path(f"chemprop_training/{args.data_path.stem}/{NOW}")
```
[the remark this answers is not in the document]
Comment (KnathanM) [reply chemprop/cli/train.py:456]: Great idea

File: chemprop/models/model.py [hunk 7]
```diff
@@ -241,39 +247,59 @@ def configure_optimizers(self):
...
-        return kwargs
+        return submodules, state_dict, hparams
+
+    @classmethod
+    def _add_metric_task_weights_to_state_dict(cls, state_dict, hparams):
+        if "metrics.0.task_weights" not in state_dict:
+            metrics = hparams["metrics"]
+            n_metrics = len(metrics) if metrics is not None else 1
+            for i_metric in range(n_metrics):
+                state_dict[f"metrics.{i_metric}.task_weights"] = torch.tensor([[1.0]])
```
[the remark this answers is not in the document]
Comment (KnathanM) [reply chemprop/models/model.py:272]: We would need to have Metric overwrite the init from LossFunction. Perhaps we could only make it a buffer for loss functions, and a normal attribute for metrics. I'd have to look into this further. See also my PR where I propose removing task_weights from the state_dict.

File: chemprop/models/model.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (shihchengli) [reply chemprop/models/model.py:170]: Let's do this in the torchmetrics PR.

File: chemprop/cli/train.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (shihchengli) [reply chemprop/cli/train.py:356]: fair point

File: chemprop/models/model.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (shihchengli) [reply chemprop/models/model.py:170]: Thanks @davidegraff @KnathanM! I think we can keep this way for now.

**Review by shihchengli — APPROVED** [review]
